zbus_address/src/address.rs

```diff
@@ -0,0 +1,233 @@
+    if value.len() != 32 || value.iter().any(|&c| !c.is_ascii_hexdigit()) {
+        return Err(Error::InvalidValue("guid".into()));
+    }
+
+    Ok(())
+}
+
+impl DBusAddr<'_> {
+    pub fn to_owned(&self) -> DBusAddr<'static> {
+        let addr = self.addr.to_string();
+        DBusAddr { addr: addr.into() }
+    }
+}
+
+impl<'a> TryFrom<String> for DBusAddr<'a> {
+    type Error = Error;
+
+    fn try_from(addr: String) -> Result<Self> {
+        Self::new(addr)
+    }
+}
+
+impl<'a> TryFrom<&'a str> for DBusAddr<'a> {
+    type Error = Error;
+
+    fn try_from(addr: &'a str) -> Result<Self> {
+        Self::new(addr)
+    }
+}
+
+impl fmt::Display for DBusAddr<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let kv = KeyValFmt::new().add("guid", self.guid());
+        let t = self.transport().map_err(|_| fmt::Error)?;
+        let kv = t.key_val_fmt_add(kv);
+        write!(f, "{t}:{kv}")?;
+        Ok(())
+    }
+}
+
+pub(super) struct KeyValIter<'a> {
+    data: &'a str,
+    next_index: usize,
+}
+
+impl<'a> KeyValIter<'a> {
+    fn new(data: &'a str) -> Self {
+        KeyValIter {
+            data,
+            next_index: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for KeyValIter<'a> {
+    type Item = (&'a str, Option<&'a str>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.next_index >= self.data.len() {
+            return None;
+        }
+
+        let mut pair = &self.data[self.next_index..];
+        if let Some(end) = pair.find(',') {
+            pair = &pair[..end];
+            self.next_index += end + 1;
+        } else {
+            self.next_index = self.data.len();
+        }
+        let mut split = pair.split('=');
+        // SAFETY: first split always returns something
+        let key = split.next().unwrap();
+        Some((key, split.next()))
+    }
+}
+
+pub(crate) trait KeyValFmtAdd {
+    fn key_val_fmt_add<'a: 'b, 'b>(&'a self, kv: KeyValFmt<'b>) -> KeyValFmt<'b>;
+}
+
+pub(crate) struct KeyValFmt<'a> {
+    fields: Vec<(Box<dyn fmt::Display + 'a>, Box<dyn Encodable + 'a>)>,
+}
+
+impl<'a> KeyValFmt<'a> {
+    fn new() -> Self {
+        Self { fields: vec![] }
+    }
+
+    pub(crate) fn add<K, V>(mut self, key: K, val: Option<V>) -> Self
+    where
+        K: fmt::Display + 'a,
+        V: Encodable + 'a,
+    {
+        if let Some(val) = val {
+            self.fields.push((Box::new(key), Box::new(val)));
+        }
+        self
+    }
+}
+
+impl fmt::Display for KeyValFmt<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut first = true;
+        for (k, v) in self.fields.iter() {
+            if !first {
+                write!(f, ",")?;
+            }
+            write!(f, "{k}=")?;
+            v.encode(f)?;
+            first = false;
+        }
+        Ok(())
+    }
+}
+
+/// A trait for objects which can be converted or resolved to one or more [`DBusAddr`] values.
+pub trait ToDBusAddrs<'a> {
```

Comment:
Why? with a trait, we can accept address arguments which are strings, or Vec<addr> etc..    This is also how std socket address are handled: https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html

Reply:
> Why? with a trait, we can accept address arguments which are strings, or Vec etc..

You can do the same with `TryFrom` impls. if `Vec` is expected, you do a `TryInto` for `AddressList` instead of `Address`. I think our case is more specific than std's.

> This is also how std socket address are handled: https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html

That API predates TryFrom/TryInto. There's a lot of API in std that's like that only for historical reasons.

Reply:
But Vec isn't expected. The implementation avoids Vec allocation... Why should we force Vec?

Reply:
My thinking was the same as here but see the new comment in that thread. Mark this also as resolved once you've read it. :)